File: src/test/java/CITestLocalProducer.java
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

import org.jlab.coda.et.*;
import org.jlab.coda.et.enums.Mode;

public class CITestLocalProducer {
    public static void main(String[] args) {


        // Parse arg(s)
        boolean noSocket = false;
        for (int i = 0; i < args.length; i++) {
            if (args[i].equalsIgnoreCase("-ns")) noSocket = true;
        }


        try {
            // Configure to open existing ET system (shared memory file) by name
            EtSystemOpenConfig config = new EtSystemOpenConfig();
            config.setEtName("/tmp/et_ci_localtest");       // ET system name (file) to connect to
            config.setHost(EtConstants.hostLocal);          // local host (since ET is local)
            config.setNetworkContactMethod(EtConstants.direct);
            System.out.println("Connecting to " + EtConstants.hostLocal);

            // Open a connection to the ET system
            EtSystem etSys = new EtSystem(config);
            etSys.open();  // establish connection to ET (throws exception if fails)

            if (etSys.usingJniLibrary()) {
                System.out.println("ET is local\n");
            }
            else {
                if(noSocket) {
                    System.out.println("No JNI library loaded and noSocket specified. Quitting!");
                    return;
                }
                System.out.println("No JNI lib: ET running in remote mode\n");
                config.setConnectRemotely(true);
                config.setTcpPort(EtConstants.serverPort);
            }

            // Attach to the GrandCentral station (for producers, GrandCentral is the source of new events)
            EtStation gc = etSys.stationNameToObject("GRAND_CENTRAL");
            EtAttachment att = etSys.attach(gc);
            
            System.out.println("Producer: waiting for consumer to attach...");
            // Wait until the consumer station "consumer_station" is created by the consumer program
            while (!etSys.stationExists("consumer_station")) {
                if (!etSys.alive()) {
                    System.err.println("ERROR: ET system not alive, exiting");
                    etSys.close();
                    return;
                }
                Thread.sleep(1000);  // sleep 1 sec and try again
                System.out.println("Still waiting for station 'consumer_station'...");
            }
            System.out.println("Producer: consumer detected, starting event transmission...");

            // Once consumer station exists, start sending 500 events
            int totalEvents = 500;
            for (int i = 1; i <= totalEvents; ++i) {
                // Request a new/free event from ET (will block until available)
                EtEvent[] newEvents = etSys.newEvents(att, Mode.SLEEP, 0, 1, 100);
                if (newEvents == null || newEvents.length == 0) {
                    System.err.printf("ERROR: et_newEvents failed at event %d\n", i);
                    break;
                }
                EtEvent ev = newEvents[0];
                ByteBuffer buf = ev.getDataBuffer();
                buf.order(ev.getByteOrder());
                // Prepare event data (message string)
                byte[] msgBytes = (("ET is great, event " + i) + "\0").getBytes(StandardCharsets.UTF_8);
                
                buf.clear();
                buf.put(msgBytes, 0, msgBytes.length);

                // Tell ET how many valid bytes are in this event
                ev.setLength(msgBytes.length);
                                
                // Send the event
                etSys.putEvents(att, new EtEvent[]{ev});

                // Copy message into event's data buffer
                // byte[] dataBuf = ev.getData();
                // int msgLen = msgBytes.length;
                // if (msgLen > dataBuf.length) msgLen = dataBuf.length;  // truncate if too long
                // System.arraycopy(msgBytes, 0, dataBuf, 0, msgLen);
                // ev.setLength(msgLen);
                // Send (put) the event to the ET system (to be delivered to consumer_station)
                // etSys.putEvents(att, new EtEvent[]{ev});
            }

            // Detach and close ET connection
            etSys.detach(att);
            etSys.close();
            System.out.println("Producer: completed sending events and disconnected.");
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: src/test/c/CITest_local_producer.c
// producer.c: Producer that waits for consumer then sends 500 events
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "et.h"

int main() {
    // Info about whole ET sytem
    et_sys_id id; et_openconfig openconfig; // For opening existing ET system
    // Info for consumer station (really only needed to check existence here)
    // et_statconfig sconfig;
    et_stat_id    stat_id_consumer;
    // Info for producer station
    et_att_id     attach_id_producer; // Attachment id for producer

    int status;

    // Open an existing ET system by name (local shared memory file)
    et_open_config_init(&openconfig);
    status = et_open(&id, "/tmp/et_ci_localtest", openconfig);
    // Like we always do after opening an ET system, we destroy the openconfig
    et_open_config_destroy(openconfig);
    if (status != ET_OK) {
        fprintf(stderr, "ERROR: Cannot open ET system (status %d)\n", status);
        exit(1);
    }

    // Attach to Grand Central station of ET system (as is typical for producer)
    et_att_id attach;
    if (et_station_attach(id, ET_GRANDCENTRAL, &attach) != ET_OK) {
        fprintf(stderr, "Error attaching to GRAND_CENTRAL\n");
        et_system_close(id);
        exit(1);
    }

    // Wait for consumer station to appear
    printf("Producer: waiting for consumer to attach...\n");
    while(!et_station_exists(id, &stat_id_consumer, "consumer_station")) {
        // If the ET system isn't alive, give up
        if (!et_alive(id)) {
            fprintf(stderr, "ERROR: ET system not alive\n");
            et_close(id);
            exit(1);
        }
        sleep(5);  // sleep 5 secs and try again
        printf("Still waiting for station '%s'...\n", "consumer_station");
    }

    // Now the consumer is connected and ready; start sending events
    printf("Producer: consumer detected, starting event transmission...\n");

    // Send 500 events
    size_t ev_size = 100;  // size of each event data buffer
    for (int i = 1; i <= 500; ++i) {
        et_event *pe;
        // Get a new/free event from ET system (blocking until available)
        status = et_event_new(id, attach_id_producer, &pe, ET_SLEEP, NULL, ev_size);
        if (status != ET_OK) {
            fprintf(stderr, "ERROR: et_event_new failed at event %d (status %d)\n", i, status);
            break;
        }
        // Fill the event buffer with our message
        char message[ev_size];
        snprintf(message, sizeof(message), "ET is great, event %d", i);
        size_t msg_len = strlen(message) + 1;  // +1 to include null terminator
        // Copy message into event data buffer
        char *ev_data = (char *) pe->pdata;   // pointer to event's data buffer
        memcpy(ev_data, message, msg_len);
        et_event_setlength(pe, msg_len);

        // Put (send) the event to the station
        status = et_event_put(id, attach_id_producer, pe);
        if (status != ET_OK) {
            fprintf(stderr, "ERROR: et_event_put failed at event %d (status %d)\n", i, status);
            break;
        }
    }

    // Detach and close ET connection
    et_station_detach(id, attach_id_producer);
    et_close(id);
    printf("Producer: completed sending events and disconnected.\n");
    return 0;
}
